gtsfm/bundle/bundle_adjustment.py

```diff
@@ -91,19 +114,48 @@ def run(
                 # i represents the camera index, and uv is the 2d measurement
                 i, uv = track.measurement(m_idx)
                 # note use of shorthand symbols C and P
-                graph.add(GeneralSFMFactorCal3Bundler(uv, measurement_noise, C(i), P(j)))
+                graph.add(
+                    GeneralSFMFactor2Cal3Bundler(
+                        uv,
+                        measurement_noise,
+                        X(i),
+                        P(j),
+                        K(self.__map_to_calibration_variable(i)),
+                    )
+                )
 
         # get all the valid camera indices, which need to be added to the graph.
         valid_camera_indices = initial_data.get_valid_camera_indices()
 
         # Add a prior on first pose. This indirectly specifies where the origin is.
         graph.push_back(
-            gtsam.PriorFactorPinholeCameraCal3Bundler(
-                C(valid_camera_indices[0]),
-                initial_data.get_camera(valid_camera_indices[0]),
-                gtsam.noiseModel.Isotropic.Sigma(PINHOLE_CAM_CAL3BUNDLER_DOF, 0.1),
+            PriorFactorPose3(
+                X(valid_camera_indices[0]),
+                initial_data.get_camera(valid_camera_indices[0]).pose(),
+                gtsam.noiseModel.Isotropic.Sigma(CAM_POSE3_DOF, CAM_POSE3_PRIOR_NOISE_SIGMA),
             )
         )
+
+        # add prior on all calibrations
+        if self._shared_calib:
+            i = valid_camera_indices[0]
```

Comment:
just have a new variable: 
camera_indices_with_prior = [0] if shared else valid_camera_indices. 

Then use the for loop with camera_indices_with_prior :)

Reply:
I implemented a similar thing. Please let me know if it works.